consensus/ethash/ethash.go

```diff
@@ -561,14 +629,40 @@ func (ethash *Ethash) SetThreads(threads int) {
 
 // Hashrate implements PoW, returning the measured rate of the search invocations
 // per second over the last minute.
+// Note the returned hashrate includes local hashrate, but also includes the total
+// hashrate of all remote miner.
 func (ethash *Ethash) Hashrate() float64 {
-	return ethash.hashrate.Rate1()
+	var (
+		total uint64
+		errCh = make(chan error)
+	)
+	// getHashrate gathers all remote miner's hashrate.
+	getHashrate := func(hashrate map[common.Hash]hashrate) {
+		for _, rate := range hashrate {
+			// this could overflow
+			total += rate.rate
+		}
+		return
+	}
```

Comment:
Lets try to convert this into a proper channel API too instead of the callback trick. It's a bit more code yes, but a lot cleaner.